### ISSUE TITLE
Potential race condition

### ISSUE THREAD

**Issue by NolanDeveloper** [coreutils/coreutils#82]
According to [Execution of signals](https://www.man7.org/linux/man-pages/man7/signal.7.html) signal handler may be called whenever there's transition between kernel and user mode. fork() is a syscall, so there should be such transition when it returns [here](https://github.com/coreutils/coreutils/blob/7b92e9fb2643371ebcecc092d8f4453a7ff23a75/src/timeout.c#L535). cleanup() function is installed for the number of termination signals before fork(). So it seems like it might be possible that cleanup() is called between the moment when fork() has finished and monitored_pid is assigned. This would make cleanup() to _exit() [here](https://github.com/coreutils/coreutils/blob/7b92e9fb2643371ebcecc092d8f4453a7ff23a75/src/timeout.c#L248). Then child process would continue running as long as it wants instead of being terminated. 

I wasn't able to reproduce this issue.

**Comment by NolanDeveloper** [comment]
I think I managed to reproduce this:

1. Add synchronization to timeout
```
diff --git a/src/timeout.c b/src/timeout.c
index 85d97c0b5..02ccbdf24 100644
--- a/src/timeout.c
+++ b/src/timeout.c
@@ -53,6 +53,7 @@
 # include <sys/prctl.h>
 #endif
 #include <sys/wait.h>
+#include <assert.h>

 #include "system.h"
 #include "cl-strtod.h"
@@ -532,6 +533,15 @@ main (int argc, char **argv)
   signal (SIGTTOU, SIG_IGN);   /* Don't stop if background child needs tty.  */
   install_sigchld ();          /* Interrupt sigsuspend() when child exits.   */

+  int fd = open("/tmp/timeout.fifo", O_RDONLY);
+  if (fd < 0) {
+      printf("open() failed.\n");
+      exit(EXIT_FAILURE);
+  }
+  char single_byte;
+  int result = read(fd, &single_byte, sizeof(single_byte));
+  assert(result == 1);
+
   monitored_pid = fork ();
   if (monitored_pid == -1)
     {
```

2. Script to reproduce issue run.sh
```
#!/bin/bash
mkfifo /tmp/timeout.fifo
for i in $(seq 10) 
do
    ./src/timeout 5 sleep 10000 & 
    pid=$!
    echo -n z >/tmp/timeout.fifo
    kill $pid
done
```

3. There is sleep process remaining
```
$ ./run.sh
$ ps
  PID TTY          TIME CMD
26165 pts/1    00:00:00 bash
26892 pts/1    00:00:00 sleep
26919 pts/1    00:00:00 ps
```

**Comment by pixelb** [comment]
Well I repro'd with:

```diff
-  monitored_pid = fork ();
+  int tpid = fork ();
+  if (tpid) sleep(10); /* delay monitor */
+  monitored_pid = tpid;
```

If you run `timeout -v -- 30 sleep 60`, and then kill the timeout process within 10s, the sleep remains

**Comment by pixelb** [comment]
Actually I think I see a related race.
If fork() fails, monitored_pid is -1.
If a signal is received before the parent process exits, then a signal is sent to pid -1, which is a all processes for which we have permission to send signals :(

**Comment by pixelb** [comment]
Fixed with commit ab4ffc850
